### PR TITLE
[SPARK-33007][SQL] Simplify named_struct + get struct field + from_json expression chain

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This proposes to simplify named_struct + get struct field + from_json expression chain from `struct(from_json.col1, from_json.col2, from_json.col3...)` to `struct(from_json)`.

### Why are the changes needed?

Simplify complex expression tree that could be produced by query optimization or user.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @HyukjinKwon @maropu @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Thank you for pining me, @viirya .

**Comment by dongjoon-hyun** [comment]
cc @sunchao

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 1]
```diff
@@ -28,10 +28,36 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
  * The optimization includes:
  * 1. JsonToStructs(StructsToJson(child)) => child.
  * 2. Prune unnecessary columns from GetStructField/GetArrayStructFields + JsonToStructs.
+ * 3  struct(from_json.col1, from_json.col2, from_json.col3...) => struct(from_json)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:31]: `3` -> `3.`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 2]
```diff
@@ -28,10 +28,36 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
...
  */
 object OptimizeJsonExprs extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case p => p.transformExpressions {
+
+      case c: CreateNamedStruct
+        if c.valExprs.forall(v => v.isInstanceOf[GetStructField] &&
+          v.asInstanceOf[GetStructField].child.isInstanceOf[JsonToStructs]) =>
+        val jsonToStructs = c.valExprs.map(_.children(0))
+        val semanticEqual = jsonToStructs.tail.forall(jsonToStructs.head.semanticEquals(_))
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:41]: Can this check be merged with L39? https://github.com/apache/spark/pull/29942/files#diff-f9d27e3c9c32aaf07bb038c779309414R39

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 3]
```diff
@@ -28,10 +28,36 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
...
+
+        // If we create struct from various fields of the same `JsonToStructs` and we don't
+        // alias field names.
+        if (semanticEqual && sameFieldName) {
+          val fromJson = jsonToStructs.head.asInstanceOf[JsonToStructs].copy(schema = c.dataType)
+          val nullFields = c.children.grouped(2).map {
+            case Seq(name, value) => Seq(name, Literal(null, value.dataType))
+          }.flatten.toSeq
+
+          If(IsNull(fromJson.child), c.copy(children = nullFields), KnownNotNull(fromJson))
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:56]: Is this related to this optimization? This looks more general to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 4]
```diff
@@ -28,10 +28,36 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
...
     case p => p.transformExpressions {
+
+      case c: CreateNamedStruct
+        if c.valExprs.forall(v => v.isInstanceOf[GetStructField] &&
+          v.asInstanceOf[GetStructField].child.isInstanceOf[JsonToStructs]) =>
+        val jsonToStructs = c.valExprs.map(_.children(0))
+        val semanticEqual = jsonToStructs.tail.forall(jsonToStructs.head.semanticEquals(_))
+        val sameFieldName = c.names.zip(c.valExprs).forall {
+          case (name, valExpr: GetStructField) =>
+            name.toString == valExpr.childSchema(valExpr.ordinal).name
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:44]: Does this work correctly if multiple values refer to the same ordinal?
```
scala> sql("""SELECT from_json('{"a":1, "b":0.8}', 'a INT, b DOUBLE') v""").createOrReplaceTempView("t")
scala> sql("select named_struct('a', t.v.a, 'a', t.v.a) from t").explain(true)
== Parsed Logical Plan ==
'Project [unresolvedalias('named_struct(a, 't.v.a, a, 't.v.a), None)]
+- 'UnresolvedRelation [t], [], false

== Analyzed Logical Plan ==
named_struct(a, v.a AS `a`, a, v.a AS `a`): struct<a:int,a:int>
Project [named_struct(a, v#128.a, a, v#128.a) AS named_struct(a, v.a AS `a`, a, v.a AS `a`)#133]
+- SubqueryAlias t
   +- Project [from_json(StructField(a,IntegerType,true), StructField(b,DoubleType,true), {"a":1, "b":0.8}, Some(Asia/Tokyo)) AS v#128]
      +- OneRowRelation

== Optimized Logical Plan ==
Project [[1,1] AS named_struct(a, v.a AS `a`, a, v.a AS `a`)#133]
+- OneRowRelation

== Physical Plan ==
*(1) Project [[1,1] AS named_struct(a, v.a AS `a`, a, v.a AS `a`)#133]
+- *(1) Scan OneRowRelation[]
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 2 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:41]: We can but L39 condition will look ugly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 3 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:56]: `JsonToStructs`'s `nullable` is true. If the input json to `JsonToStructs` is null, we will get a null output. But `CreateNamedStruct`'s `nullable` is false, so here we need to keep nullability unchanged by wrapping them with a `If(IsNull...)`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:44]: Oh, this is kind of corner case. We cannot simplify this case. Added test case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala]: fixed. thanks.

**Comment by viirya** [comment]
Thanks for quick response. Addressed the comments.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 3 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:60]: Ah, ok. I see.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 2 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:41]: hm I see. I noticed that it looped `c.valExprs` `{3 x len(c.valExprs)}` times to check the condition. Minor optimization though, I thought it would be nice if it could stop early if the condition not met.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:41]: Ok, let me change it and see how it looks like.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala]: Moved the condition to top.

**Comment by viirya** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 5]
```diff
@@ -28,10 +28,43 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
  * The optimization includes:
  * 1. JsonToStructs(StructsToJson(child)) => child.
  * 2. Prune unnecessary columns from GetStructField/GetArrayStructFields + JsonToStructs.
+ * 3. struct(from_json.col1, from_json.col2, from_json.col3...) => struct(from_json)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:31]: It's a nit but can we match the comment style? 1. and 2. used the expression class names but 3. used function-ish name.

BTW, `sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst` is a private package that doesn't generate Javadoc. So it should be safe to use `[[...]]` as we want (because using `[[..]]` in Scaladoc causes some problems sometimes when it's converted into Javadoc in some cases such as `trait`s).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 6]
```diff
@@ -28,10 +28,43 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
...
+          if c.valExprs.forall { v =>
+            v.isInstanceOf[GetStructField] &&
+              v.asInstanceOf[GetStructField].child.isInstanceOf[JsonToStructs] &&
+              v.children.head.semanticEquals(c.valExprs.head.children.head)
+          } =>
+        val jsonToStructs = c.valExprs.map(_.children.head)
+        val sameFieldName = c.names.zip(c.valExprs).forall {
+          case (name, valExpr: GetStructField) =>
+            name.toString == valExpr.childSchema(valExpr.ordinal).name
+          case (_, _) => false
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:48]: nit: `case (_, _) =>` -> `case _ =>`

**Comment by HyukjinKwon** [comment]
@viirya just to clarify, is it to avoid calling the same `from_json` multiple times? How does it relate to SPARK-32939 and  SPARK-32943?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 5 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:31]: Yeah, changed the comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 6 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala]: fixed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 7]
```diff
@@ -28,10 +28,45 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
  * The optimization includes:
  * 1. JsonToStructs(StructsToJson(child)) => child.
  * 2. Prune unnecessary columns from GetStructField/GetArrayStructFields + JsonToStructs.
+ * 3. CreateNamedStruct(JsonToStructs(json).col1, JsonToStructs(json).col2, ...) =>
+ *      CreateNamedStruct(JsonToStructs(json)) if JsonToStructs(json) is shared among all
+ *      fields of CreateNamedStruct.
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:33]: For a fresh eye with no context this is still a bit confusing - does the list `col1`, `col2` etc have to represent all columns in the `json` struct?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprsSuite.scala [hunk 8]
```diff
@@ -199,4 +199,51 @@ class OptimizeJsonExprsSuite extends PlanTest with ExpressionEvalHelper {
...
+    comparePlans(optimized1, expected1)
+
+    // Skip it if `namedStruct` aliases field name.
+    val field1 = StructType.fromDDL("a int")
+    val field2 = StructType.fromDDL("b int")
+    val query2 = testRelation2
+      .select(namedStruct(
+        "a1", GetStructField(JsonToStructs(schema, options, 'json), 0),
+        "b", GetStructField(JsonToStructs(schema, options, 'json), 1)).as("struct"))
+    val optimized2 = Optimizer.execute(query2.analyze)
```
Comment (sunchao) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprsSuite.scala:229]: seems this is a bit repetitive - perhaps we can create a util method for the comparison? we can test evaluation in the method too.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 9]
```diff
@@ -28,10 +28,45 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
...
+            name.toString == valExpr.childSchema(valExpr.ordinal).name
+          case _ => false
+        }
+
+        // Although `CreateNamedStruct` allows duplicated field names, e.g. "a int, a int",
+        // `JsonToStructs` does not support parsing json with duplicated field names.
+        val duplicateFields = c.names.map(_.toString).distinct.length != c.names.length
+
+        // If we create struct from various fields of the same `JsonToStructs` and we don't
+        // alias field names and there is not duplicated fields in the struct.
```
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:58]: nit: "there is not duplicated fields" -> "there is no duplicated field"

**Comment by sunchao** [comment]
Thanks @dongjoon-hyun for pinging and left some comments @viirya (sorry some comments are stale so pls ignore them).

**Comment by viirya** [comment]
> @viirya just to clarify, is it to avoid calling the same `from_json` multiple times? How does it relate to [SPARK-32939](https://issues.apache.org/jira/browse/SPARK-32939) and [SPARK-32943](https://issues.apache.org/jira/browse/SPARK-32943)?

This patch targets specifically for a special pattern `CreateNamedStruct` + multiple `GetStructField` of same `JsonToStructs`, it could be produced by the optimizer or by users manually.

Sometimes the query optimizer can optimize a query to have many duplicated expressions e.g. `JsonToStructs`. This is SPARK-32943 wants to fix. It targets a broader problem.

For SPARK-32939, because it was not reported by me, some details I might not get from its description. We don't de-duplicate expressions in whole-stage codegen overall (but only in specified operator). If we disable whole-stage codegen, interpreted Project will de-duplicate expressions for some cases (`GenerateUnsafeProjection`), but not always (we could also fallback to `InterpretedUnsafeProjection` possibly). For specified expressions like `CaseWhen`, we have a chance to de-duplicate the condition expressions, if we want.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 7 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:33]: No, it could be part of the json struct. In the case, we will prune unnecessary columns in `JsonToStructs`.

**Comment by maropu** [comment]
No more comment and it looks okay.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprsSuite.scala (hunk 8 above)
Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprsSuite.scala]: Ok.

**Comment by viirya** [comment]
Thanks @maropu

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 9 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala]: fixed.

**Comment by HyukjinKwon** [comment]
No more comments from me too. I am okay with this given that we have a plan for related tickets (https://github.com/apache/spark/pull/29942#issuecomment-703851082).

**Comment by viirya** [comment]
Thanks @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 10]
```diff
@@ -28,10 +28,45 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
...
+            name.toString == valExpr.childSchema(valExpr.ordinal).name
+          case _ => false
+        }
+
+        // Although `CreateNamedStruct` allows duplicated field names, e.g. "a int, a int",
+        // `JsonToStructs` does not support parsing json with duplicated field names.
+        val duplicateFields = c.names.map(_.toString).distinct.length != c.names.length
+
+        // If we create struct from various fields of the same `JsonToStructs` and we don't
+        // alias field names and there is not duplicated field in the struct.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:58]: `not` -> `no`?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:58]: oh, will fix it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala (hunk 7 above)
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:33]: Do we have a test coverage for this where accessing all columns?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 11]
```diff
@@ -28,10 +28,45 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
...
+        val jsonToStructs = c.valExprs.map(_.children.head)
+        val sameFieldName = c.names.zip(c.valExprs).forall {
+          case (name, valExpr: GetStructField) =>
+            name.toString == valExpr.childSchema(valExpr.ordinal).name
+          case _ => false
+        }
+
+        // Although `CreateNamedStruct` allows duplicated field names, e.g. "a int, a int",
+        // `JsonToStructs` does not support parsing json with duplicated field names.
+        val duplicateFields = c.names.map(_.toString).distinct.length != c.names.length
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:55]: `c.names.map(_.toString).distinct.length` seems to ignore case-insensitive situation. Could you add a test case for column `A` and `a` in case-insensitive mode?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala [hunk 12]
```diff
@@ -28,10 +28,45 @@ import org.apache.spark.sql.types.{ArrayType, StructType}
  * The optimization includes:
  * 1. JsonToStructs(StructsToJson(child)) => child.
  * 2. Prune unnecessary columns from GetStructField/GetArrayStructFields + JsonToStructs.
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:30]: @viirya . It would be great if we describe `If(IsNull(json),  nullStruct, KnownNotNull(JsonToStructs(prunedSchema1, options, json, ..))` pattern here because technically it's a new optimized form but not a simplified one. The existing (1), (2), (3) of this PR doesn't cover it.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeJsonExprs.scala:30]: Ok, revised the comment here. Thanks.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @viirya and all.
Merged to master for Apache Spark 3.1.0 on December 2020.

**Comment by viirya** [comment]
Thanks!
